cpp/dolfinx/fem/Expression.h

```diff
@@ -28,29 +28,45 @@ class Constant;
 /// at quadrature points in all cells. This evaluated gradient can then
 /// be used as input in to a non-FEniCS function that calculates a
 /// material constitutive model.
+
 template <typename T>
 class Expression
 {
 public:
   /// Create an Expression
   ///
+  /// Users should prefer the create_expression factory functions
+  ///
   /// @param[in] coefficients Coefficients in the Expression
   /// @param[in] constants Constants in the Expression
   /// @param[in] mesh
-  /// @param[in] X points on reference cell, shape=(num_points, tdim)
+  /// @param[in] X points on reference cell, number of points rows and
+  /// tdim cols
   /// @param[in] fn function for tabulating expression
-  /// @param[in] value_size size of expression evaluated at single point
+  /// @param[in] value_shape shape of expression evaluated at single point
+  /// @param[in] argument_function_space Function space for Argument
   Expression(
       const std::vector<std::shared_ptr<const Function<T>>>& coefficients,
       const std::vector<std::shared_ptr<const Constant<T>>>& constants,
-      const std::shared_ptr<const mesh::Mesh>& mesh,
       const xt::xtensor<double, 2>& X,
-      const std::function<void(T*, const T*, const T*, const double*)> fn,
-      std::size_t value_size)
-      : _coefficients(coefficients), _constants(constants), _mesh(mesh), _x(X),
-        _fn(fn), _value_size(value_size)
+      const std::function<void(T*, const T*, const T*, const double*,
+                               const int*, const uint8_t*)>
+          fn,
+      const std::vector<int>& value_shape,
+      const std::shared_ptr<const mesh::Mesh> mesh = nullptr,
```

Comment:
It's unusual to pass by value with `const`. Normally I'd expect non-const, or `const` and pass by reference.

Reply:
I think we should use non-const shared_ptr by value. Const ptr by reference means it only "might" retain refcount, https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam-const, but we know we do certainly retain refcount.

I'll use const ptr by reference here to be consistent with the rest codebase.

Reply:
I think in our use, const ref does mean that an object will be created/returned that will bump the ref count. Otherwise we would pass a plain reference.

Reply:
Const ref doesn't tell you if the shared_ptr is later assigned (constructed) into member variable. I think thats what guideline is trying to say. But since we know we always do `_mesh(mesh)` then we should pass ptr by value, so interface makes that clear.